File: interpreter/python/job_manager.py
# Created by Sezer BOZKIR<admin@sezerbozkir.com> at 20.12.2021
from developer_expression import DeveloperExpression
from team_lead_expression import TeamLeadExpression
from accounting_expression import AccountingExpression
from context import Context


class JobManager:
    @staticmethod
    def create_expression_tree(formula: str):
        tree = []
        for role in formula:
            if role.lower() == 'g':
                tree.append(DeveloperExpression())
            elif role.lower() == 'm':
                tree.append(AccountingExpression())
            elif role.lower() == 'l':
                tree.append(TeamLeadExpression())
            else:
                print(role)
                raise Exception("Unexpected role!")
        return tree

    def run_expression(self, context: Context):
        for expression in self.create_expression_tree(context.formula):
            expression.interpret(context)

File: interpreter/typescript/job-manager.ts
import {Context} from "./context";
import {IJobExpression} from "./job-expression.interface";
import {DeveloperExpression} from "./developer-expression";
import {AccountingExpression} from "./accounting-expression";
import {TeamLeadExpression} from "./team-lead-expression";

export class JobManager {
  public createExpressionTree(formula: string): IJobExpression[] {
    const tree: IJobExpression[] = [];

    for(const role of formula) {
      switch (role) {
        case 'G':
          tree.push(new DeveloperExpression())
          break;
        case 'M':
          tree.push(new AccountingExpression())
          break;
        case 'L':
          tree.push(new TeamLeadExpression())
          break;
        default:
          throw new Error('Unexpected role!');
      }
    }

    return tree;
  }

  public runExpression(context: Context): void {
    for (let expression of this.createExpressionTree(context.formula)) {
      expression.interpret(context);
    }
  }
}
